---
title: "Class activity - small sample t tests"
author: "ks"
date: "6/10/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
if(!require("genefilter")){BiocManager::install("genefilter")} 
library(genefilter)
library(tidyverse)
library(matrixStats)
```

## 2-sample t-test: Single test

Let's estimate the significance level for the 2-sample t-test.  

Simulation settings:  
n1=n2=3 is the sample size in each group  
Each measurement ~ N(6,sd=0.5)  
Simulate 1,000 replicates (rows)

How many tests reject the null hypothesis?  I can use the rowttests() function in R to do this quickly for the 10000 replicates.

```{r sim1gene}
n1 <- 3
n  <- 2*n1
group <- factor(rep(c("Group1","Group2"),each=n1))
                
set.seed(40)
simdat <- replicate(n,rnorm(n = 10000, mean = 6, sd = 0.5))
dim(simdat)
tt <- genefilter::rowttests(simdat,group)
mean(tt$p.value<0.05)
```



## 2-sample t-test: 20,000 genes with different means and variances

We want to test for differential expression in two treatment groups (treatment vs control), each run in triplicate. How does the t-test perform in small sample sizes?

Let's run a simulation to see how the t-test behaves in small sample sizes of 3 vs 3, when there is no treatment effect (null hypothesis).  

In gene expression studies, the mean and variances of the genes we study are not all the same. Let's use the mean gene expression profile and estimates of standard deviation from the data set we've been using in class (jbcdat.rda). First, we load the data from our directory.

**Simulated data**

```{r ReadData}
load("data/JBC 2012/jbcdat.rda")
```

Compute the mean and standard deviations for 20,000 features from this data set. 

```{r summarize_data}
set.seed(22)
idx   <-sample(c(1:47231),20000)
avg_e <- rowMeans(jbcdat$E[idx,])
sd_e  <- matrixStats::rowSds(jbcdat$E[idx,])
```

1 observation has 20000 genes ~ N(mean_vec, sd_vec), where mean_vec and sd_vec are the vectors of 20000 means and 20000 standard deviations we computed from our data set. 

To evaluate our test in small sample sizes we want to simulate data under the null for 6 independent observations (3 per group).  

```{r simdat}
ng <- length(avg_e)
set.seed(40)
simdat <- replicate(6,rnorm(n = ng, mean = avg_e, sd = sd_e))
dim(simdat)
```


Without loss of generality, assume the first 3 columns are observations from group 1 and the last 3 columns are from group 2. The t-test is computed from the group averages and sds. Compute these now.
```{r sumsim}
gp1_avg <- 
gp2_avg <- 

gp1_sd <- 
gp2_sd <- 
```

What do you expect to see when you plot gp1_avg vs gp2_avg?

Let's check:
```{r smoothscatter2}  
  smoothScatter(gp1_avg,gp2_avg)
  lines(lowess(gp1_avg,gp2_avg),col=2)
  abline(0,1)
```

What do you expect to see when you plot gp1_sd vs gp2_sd?

Let's check:
```{r smoothscatter3}  
  smoothScatter(gp1_sd,gp2_sd)
  lines(lowess(gp1_sd,gp2_sd),col=2)
  abline(0,1)
```

What does this show?  

Let's zoom in on the range with sds less than 0.3. 
```{r smoothscatter4}  
  smoothScatter(gp1_sd,gp2_sd,xlim=c(0,0.3),ylim=c(0,1))
  lines(lowess(gp1_sd,gp2_sd),col=2,lwd=2)
  abline(0,1)
```

The t-statistic uses the pooled estimate of standard deviation, so let's compute the estimate of the pooled standard deviation and compare it to the true simulation value (sd_e). 
Display a smooth scatter diagram of the pooled standard deviation vs the true value, and limit the range of true values to 0 and 0.3. This will zoom in on the area of the plot with lots of small standard deviations.
```{r pooledvstruesd}  
pooled_sd <- sqrt( (gp1_sd^2*2 + gp2_sd^2*2)/4)
  smoothScatter(sd_e,pooled_sd,xlim=c(0,0.3))
  lines(lowess(sd_e,pooled_sd),col=2)
  abline(0,1)
```

What do you see? How would this affect the t-statistics?

Compute the t-tests and plot the absolute value of the t-statistics vs the pooled sd to find out.  The command for absolute value is abs(x).
```{r tstats}  
gp <- factor(rep(c("group1","group2"),each=3))
tt <- genefilter::rowttests(simdat,gp)
#head(tt,n=3)
smoothScatter(pooled_sd,     )
lines(lowess(pooled_sd,     ),col=2,lwd=2)
```

```{r sessionInfo}
sessionInfo()
```

