---
title: "Differential Expression Supplement"
author: "ks"
date: "06/10/2020"
output: html_document
---

# {.tabset}

## Data 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
if (!require(genefilter)) BiocManager::install("genefilter")
library(genefilter)
library(limma)
library(tidyverse)
```


We will analyze the prostate cancer cell line data we've used in class. The treatments are:  

treatment 1: siNS, sip300, siCBP 

treatment 2:  0 hr,  16 hrs
     
All cells are treated with androgen at 0 hrs. The genes that change expression between 0 and 16 hours are genes that are regulated by androgen. These might vary based on the different treatments siNS (control), sip300, and siCBP. Let's compare average gene expression between 0 and 16 hrs for the control reaction.

```{r ReadData}
jbcdir=c("data/JBC 2012")
load(file.path(jbcdir,"jbcdat.rda"))
```

Here we subset to the control experiments at 0 and 16 hours.

```{r subset-data}
geoid <- rownames(jbcdat$targets)[jbcdat$targets$treatment == "siNS"]
subsetEobj <- function(dataobj,geonames){
  subobj <-NULL
     subobj$targets <- dataobj$targets[geonames,]
     subobj$E       <- dataobj$E[,geonames]
     subobj$genes   <- dataobj$genes
     subobj
}

siNSobj <- subsetEobj(jbcdat,geoid)
# check the annotation and expression are both subset identically
identical(rownames(siNSobj$targets),colnames(siNSobj$E))
```

## Mod t calcs

I will start with computing the moderated t-tests, and then check the calculations of different elements of the t-test and moderated t-test. 
```{r efit}
design <- model.matrix(~factor(siNSobj$targets$hour))
fit  <- limma::lmFit(siNSobj$E,design) 
efit <- limma::eBayes(fit)
```

First we can compute the difference in means between the 2 groups (log fold-change) and show this equals the coefficient from the regression model.

```{r logfc}
mn.0h <- rowMeans(siNSobj$E[,siNSobj$targets$hour=="0hr"])
mn.16h <- rowMeans(siNSobj$E[,siNSobj$targets$hour=="16hr"])
logFC <- mn.16h-mn.0h
calcfc <- cbind.data.frame(mn0h = mn.0h[1:3],
                 logfc =logFC[1:3])
calcfc
fit$coefficients[1:3,]
```

Now we show the calculation for the pooled standard deviation, and compare it to the value estimated from the regression.

```{r scalc}
v1 <- rowVars(siNSobj$E[,siNSobj$targets$hour=="0hr"])
v2 <- rowVars(siNSobj$E[,siNSobj$targets$hour=="16hr"])
poolSds <- sqrt( (v1*3 + v2*3)/6)
NSsds <- fit$sigma 
summary(poolSds-NSsds)
```

These are all within rounding error of 0. 

Compute the 2-sample t-tests.
```{r tstat}
fit$coefficients[1:3,2]/fit$sigma[1:3]/sqrt(1/4+1/4)
rt <- rowttests(siNSobj$E,factor(siNSobj$targets$hour))
rt$statistic[1:3]
```

The rowttests function selected 16hr as baseline instead of 0hr, so the signs are different, but otherwise the results are the same.

Next is the calculation of the moderated variance estimate.

```{r mods}
mysds.post <- (efit$df.prior*efit$s2.prior + 
                  efit$df.residual*fit$sigma^2)/(efit$df.prior + efit$df.residual)
mysds.post[1:3]
efit$s2.post[1:3]
```

And lastly, calculating the moderated t-test.

```{r modtstat}
efit$coefficients[1:3,2]/sqrt(efit$s2.post[1:3])/sqrt(1/4 + 1/4)
efit$t[1:3,2]
```


## Hist logSd


```{r histsd,echo=FALSE}
NSsds <- fit$sigma 
hist(NSsds,breaks=20)
```
  
```{r histlogsd,echo=FALSE}
hist(log(NSsds),breaks=20)
```

## Simulate Null

Here we're going to simulate data under the null to estimate the false-positive rate from a 2 sample t-test as a function of the sample standard deviation when the sample size is small (n1 = n2 = 4).

```{r simdat}
ng <- length(NSsds)
set.seed(40)
simdat <- replicate(8,rnorm(n = ng, mean = 0, sd = NSsds))
```

```{r ttests}
tt=genefilter::rowttests(simdat,
                         factor(rep(c("gp1","gp2"),each = 4)))
tt[1,]
```

```{r smscatt_test}  
ttsd <- cbind.data.frame(tt,
                        pooled_sd = tt$dm/tt$statistic*sqrt(1/4+1/4))
```

```{r plot_tstat_vs_pooledse}
smoothScatter(ttsd$pooled_sd,ttsd$statistic,xlab="Pooled sd",
              ylab="t stat",xlim=c(0,.6),
              main=" t stat vs sd")
abline(qt(0.025,4),0,col=2,lwd=2)
abline(qt(0.975,4),0,col=2,lwd=2)
```


```{r tvsds,echo=FALSE}
smoothScatter(ttsd$pooled_sd,abs(ttsd$statistic),xlab="Pooled sd",
              ylab="Abs(t stat)",
              main=" Abs(t stat) vs sd")
abline(qt(0.975,4),0,col=2,lwd=2)
```

## Empirical Error

```{r empiricalerror}
library(tidyverse)
ttfp <- 
ttsd %>% 
  mutate(poolsd_decile = ntile(pooled_sd, 10))
  
sumquant <- 
  ttfp %>%
  group_by(poolsd_decile) %>%
  summarize(emp.pval_05 = mean(p.value<0.05),
            emp.pval_01 = mean(p.value<0.01))
  sumquant <- cbind.data.frame(sumquant,
      quantpldsd = quantile(ttfp$pooled_sd,probs=seq(0.05,0.95,.1)))
```

```{r plot}
plot(sumquant$quantpldsd,sumquant$emp.pval_05,
     type="l", ylim=c(0,0.15),
     xlab = "Pooled SD", 
     ylab = "Pr(|T| > t)",
     main = "Empirical false-positive prob.")
text(0.05,0.13,expression(paste(alpha ,"= 0.05")),cex=1.3)
lines(sumquant$quantpldsd,sumquant$emp.pval_01,
      lty=2)
text(0.06,.035,expression(paste(alpha ,"= 0.01")),cex=1.3)
```

## Empirical Error for Moderated t

```{r modT}
design <- model.matrix(~factor(rep(c("gp1","gp2"),each = 4)))
fit  <- limma::lmFit(simdat,design)
efit <- limma::eBayes(fit)

modt <- data.frame(statistic = efit$t[,2],
                      pooled_sd = efit$sigma,
                   p.value = efit$p.value[,2])
```


```{r empiricalerrormodt}
modtfp <- 
modt %>% 
  mutate(poolsd_decile = ntile(pooled_sd, 10))
  
sumquant <- 
  modtfp %>%
  group_by(poolsd_decile) %>%
  summarize(emp.pval_05 = mean(p.value<0.05),
            emp.pval_01 = mean(p.value<0.01))
  sumquant <- cbind.data.frame(sumquant,
      quantpldsd = quantile(modtfp$pooled_sd,probs=seq(0.05,0.95,.1)))
```

```{r plotempiricalfpr}
plot(sumquant$quantpldsd,sumquant$emp.pval_05,
     type="l", ylim=c(0,0.15),
     xlab = "Pooled SD", 
     ylab = "Pr(|mod T| > t)",
     main = "Empirical false-positive prob.")
text(0.1,0.08,expression(paste(alpha ,"= 0.05")),cex=1.3)
lines(sumquant$quantpldsd,sumquant$emp.pval_01,
      lty=2)
text(0.1,.03,expression(paste(alpha ,"= 0.01")),cex=1.3)
abline(0.01,0,col=2)
abline(0.05,0,col=2)
```

## .ppt Fig

```{r panelplot}
par(mfrow=c(2,2))
hist(NSsds,breaks=20)

smoothScatter(ttsd$pooled_sd,ttsd$statistic,xlab="Pooled sd",
              ylab="t stat",xlim=c(0,.6),
              main=" t stat vs sd")
abline(qt(0.025,4),0,col=2,lwd=2)
abline(qt(0.975,4),0,col=2,lwd=2)

smoothScatter(ttsd$pooled_sd,abs(ttsd$statistic),xlab="Pooled sd",
              ylab="abs(t stat)",
              main=" abs(t stat) vs sd")
abline(qt(0.975,4),0,col=2,lwd=2)


sumquant <- 
  ttfp %>%
  group_by(poolsd_decile) %>%
  summarize(emp.pval_05 = mean(p.value<0.05),
            emp.pval_01 = mean(p.value<0.01))
  sumquant <- cbind.data.frame(sumquant,
      quantpldsd = quantile(ttfp$pooled_sd,probs=seq(0.05,0.95,.1)))

plot(sumquant$quantpldsd,sumquant$emp.pval_05,
     type="l", ylim=c(0,0.15),
     xlab = "Pooled SD", 
     ylab = "Pr(|T| > t)",
     main = "Empirical 
     false-positive prob.")
text(0.07,0.13,expression(paste(alpha ,"= 0.05")),cex=1)
lines(sumquant$quantpldsd,sumquant$emp.pval_01,
      lty=2)
text(0.085,.025,expression(paste(alpha ,"= 0.01")),cex=1)
abline(0.05,0,col=2)
```

## SessionInfo

```{r sessionInfo}
sessionInfo()
```
